Remove well-alignedness assumption from pointer dereferencing

tautschnig opened the pull request:
With void* pointers, pointer dereferencing would permit any type for the
dereferenced object. For arrays, alignment at the root level was
implicitly assumed, not taking into account that there may be a
well-aligned pointer to a nested element.

For the regression test we got `arr[11 / 7]` when dereferencing `(void
*)ptr`, where `ptr` was `(char *)&arr[0] + 11` (and the size of `arr`'s
elements is 7 bytes).

Fixes: #7036

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Remove well-alignedness assumption from pointer dereferencing

With void* pointers, pointer dereferencing would permit any type for the
dereferenced object. For arrays, alignment at the root level was
implicitly assumed, not taking into account that there may be a
well-aligned pointer to a nested element.

For the regression test we got `arr[11 / 7]` when dereferencing `(void
*)ptr`, where `ptr` was `(char *)&arr[0] + 11` (and the size of `arr`'s
elements is 7 bytes).

Fixes: #7036

jimgrundy commented:
This PR should resolve a soundness issue. Can we prioritize review?

kroening commented:
If removing a typecast fixes a soundness issue, then you haven't fixed the soundness issue, you've just avoided it

tautschnig commented:
> If removing a typecast fixes a soundness issue, then you haven't fixed the soundness issue, you've just avoided it

@kroening: I'm not really sure whether there is a bug in dereferencing or whether it is doing the right thing for `void*` pointers. The approach taking here indeed avoids that problem. See my comment in the commit message: "For the regression test we got `arr[11 / 7]` when dereferencing `(void *)ptr`, where `ptr` was `(char *)&arr[0] + 11` (and the size of `arr`'s elements is 7 bytes)." That expression loses an offset of 4 bytes, which we could fix by adding pointer arithmetic (over `char *`) during dereferencing. But then that's only as obvious when the offset is constant. Guidance would be most appreciated.

jimgrundy commented:
I thought the C Standard forbade dereferencing void pointers (pointers to an incomplete type). So, arguably the right thing for us to do would be to error on these. Would that address this?

peterschrammel commented:
> the C Standard forbade dereferencing void pointers (pointers to an incomplete type).

Afaics, the `void*` cast is only introduced because of the declaration of the array intrinsics. Declaring them with `char*` makes more sense than `void*`.

kroening commented:
`void *` is the right choice. See `memcpy` or similar library functions.
It's not right to just avoid the issue, this can only be a temporary hack fix.

kroening commented:
How about teaching `process_array_expr` to handle typecasts!

tautschnig committed:
Pointer dereferencing cleanup: Make try_add_offset_to_indices file-local

This function does not require access to any state, and is really not
generally usable or useful. Also, make explicit why it would not work
for type casts.

tautschnig committed:
Pointer dereferencing: code cleanup in build_reference_to

Avoid starting with an uninitialised expression, instead declare it at
the point of definition.

tautschnig commented:
> How about teaching `process_array_expr` to handle typecasts!

Thank you for pushing back @kroening. I have now instead fixed the (unsound) well-alignedness asumption in pointer dereferencing. Said code happily deals with the case of type casts, and so it was neither necessary to tweak `process_array_expr` nor was it necessary to change the signature of the array functions in the C front-end.
